torch/_utils.py

```diff
@@ -393,3 +393,59 @@ def reraise(self):
             # (https://bugs.python.org/issue2651), so we work around it.
             msg = KeyErrorMessage(msg)
         raise self.exc_type(msg)
+
+def _get_available_device_type():
+    if torch.cuda.is_available():
+        return "cuda"
+    # add more available device types here
+    return None
+
+def _get_device_attr(get_member):
+    device_type = _get_available_device_type()
+    if device_type.lower() == "cuda":
+        return get_member(torch.cuda)
+    # add more available device types here
+    return None
+
+def _get_current_device_index():
+    # current device index
+    return _get_device_attr(lambda m: m.current_device())
+
+def _get_all_devices_index():
+    # all device index
+    return _get_device_attr(lambda m: list(range(m.device_count())))
+
+def _get_devices_properties(device_ids):
+    # all device properties
+    return [_get_device_attr(lambda m: m.get_device_properties(i)) for i in device_ids]
+
+def _get_device_index(device, optional=False):
+    r"""Gets the device index from :attr:`device`, which can be a torch.device
+    object, a Python integer, or ``None``.
+
+    If :attr:`device` is a torch.device object, returns the device index if it
+    has index. Note that for a device without a specified index,
+    i.e., ``torch.device('xxx')``, this will return the current default supported
+    device if :attr:`optional` is ``True``.
+
+    If :attr:`device` is a Python integer, it is returned as is.
+
+    If :attr:`device` is ``None``, this will return the current default supported
+    device if :attr:`optional` is ``True``.
+    """
+    if isinstance(device, torch._six.string_classes):
+        device = torch.device(device)
+    if isinstance(device, torch.device):
+        if device.type == 'cpu':
+            raise ValueError('Expected a non cpu device, but got: {}'.format(device))
+        device_idx = device.index
+    else:
+        device_idx = device
+    if device_idx is None:
+        if optional:
+            device_idx = _get_current_device_index()
+        else:
+            raise ValueError('Expected a torch.device with a specified index '
+                             'or an integer, but got: '.format(device))
```

Comment:
Missing a `{}` with the provided device? And this also means, this code pass is not tested. Let's add a test for it.

Reply:
The _get_device_index is moved from torch/cuda/_utils.
Make the torch/cuda/_utils to use the function in the torch/_utils.
We can leverage the old CUDA test case.